yt/frontends/amrvac/tests/test_outputs.py

```diff
@@ -21,118 +26,113 @@
-    grids = ds.index.grids
-    assert ds.index.max_level == 2
-    for g in grids:
-        assert isinstance(g, AMRVACGrid)
-        assert isinstance(g.LeftEdge, YTArray)
-        assert isinstance(g.RightEdge, YTArray)
-        assert isinstance(g.ActiveDimensions, np.ndarray)
-        assert isinstance(g.Level, (np.int32, np.int64, int))
-
-
-@requires_ds(blastwave_polar_2D)
-def test_bw_polar_2d():
-    ds = data_dir_load(blastwave_polar_2D)
-    for test in small_patch_amr(ds, _get_fields_to_check(ds)):
-        test_bw_polar_2d.__name__ = test.description
-        yield test
-
-
-@requires_ds(blastwave_cartesian_3D)
-def test_blastwave_cartesian_3D():
-    ds = data_dir_load(blastwave_cartesian_3D)
-    for test in small_patch_amr(ds, _get_fields_to_check(ds)):
-        test_blastwave_cartesian_3D.__name__ = test.description
-        yield test
-
-
-@requires_ds(blastwave_spherical_2D)
-def test_blastwave_spherical_2D():
-    ds = data_dir_load(blastwave_spherical_2D)
-    for test in small_patch_amr(ds, _get_fields_to_check(ds)):
-        test_blastwave_spherical_2D.__name__ = test.description
-        yield test
-
-
-@requires_ds(blastwave_cylindrical_3D)
-def test_blastwave_cylindrical_3D():
-    ds = data_dir_load(blastwave_cylindrical_3D)
-    for test in small_patch_amr(ds, _get_fields_to_check(ds)):
-        test_blastwave_cylindrical_3D.__name__ = test.description
-        yield test
-
-
-@requires_ds(khi_cartesian_2D)
-def test_khi_cartesian_2D():
-    ds = data_dir_load(khi_cartesian_2D)
-    for test in small_patch_amr(ds, _get_fields_to_check(ds)):
-        test_khi_cartesian_2D.__name__ = test.description
-        yield test
-
-
-@requires_ds(khi_cartesian_3D)
-def test_khi_cartesian_3D():
-    ds = data_dir_load(khi_cartesian_3D)
-    for test in small_patch_amr(ds, _get_fields_to_check(ds)):
-        test_khi_cartesian_3D.__name__ = test.description
-        yield test
-
-
-@requires_ds(jet_cylindrical_25D)
-def test_jet_cylindrical_25D():
-    ds = data_dir_load(jet_cylindrical_25D)
-    for test in small_patch_amr(ds, _get_fields_to_check(ds)):
-        test_jet_cylindrical_25D.__name__ = test.description
-        yield test
-
-
-@requires_ds(riemann_cartesian_175D)
-def test_riemann_cartesian_175D():
-    ds = data_dir_load(riemann_cartesian_175D)
-    for test in small_patch_amr(ds, _get_fields_to_check(ds)):
-        test_riemann_cartesian_175D.__name__ = test.description
-        yield test
-
-
-@requires_ds(rmi_cartesian_dust_2D)
-def test_rmi_cartesian_dust_2D():
-    # dataset with dust fields
-    ds = data_dir_load(rmi_cartesian_dust_2D)
-    for test in small_patch_amr(ds, _get_fields_to_check(ds)):
-        test_rmi_cartesian_dust_2D.__name__ = test.description
-        yield test
+    # return [fields, field_ids]
+    return [ds, fields]
+
+
+def get_pairs():
+    f_list = [_get_fields_to_check(fname) for fname in ds_list]
+    pairs = [(i[0], f) for i in f_list for f in i[1]]
```

Comment:
This is very difficult to read to me, to the point I'm not even sure I can suggest a better version.

Reply:
@neutrinoceros It is hard to read, I agree. Basically, I needed `(ds, field)` pairs and I couldn't find an easy way to "distribute" the dataset object to each field required by that dataset. I looked at itertools, but didn't see anything (though it's possible I missed the solution). Basically, in the original version, each function was calling the exact same code, but for a different dataset (e.g., `test_blastwave_cartesian_3D` and `test_blastwave_spherical_2D` have the exact same code). The goal here was to streamline this into one function that could be parametrized by pytest to work for each dataset. The issue was that, as said above, I needed to pair each dataset with each of the fields corresponding to that, hence the gross list comprehension.

Reply:
You may want to take a look a `more_itertools`, though I can't recall off the top of my head if they have something similar. It's not part of the standard library, but it's already part of a yt-compatible env thanks to transitive dependencies.
Even if there's no better approach we can think of, I find it confusing that `i` iterates over `f_list` while `f` iterates over `i[1]`. Also I don't know exactly what `f` is supposed to represent. I'm guessing a clearer refactor would look something like
```suggestion
    pairs = [(ftype, fname) for (ftype, fnames) in f_list for f in fnames]
```